File: crates/sorcery-engine/tests/gauntlet.rs
use serde_json::{Value, json};
use sorcery_engine::batch::BatchJob;
use sorcery_engine::canonical::{canonical_json, identity_hash};
use sorcery_engine::gauntlet::{GauntletOrientation, GauntletPair, GauntletReport, run_gauntlet};
use sorcery_engine::policy::{PolicySnapshot, parse_policy_snapshot};
use sorcery_engine::synthetic::synthetic_demo_manifest_json;

const HASH_B: &str = "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb";
const HASH_C: &str = "sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc";

fn swapped_manifest(manifest: &str) -> String {
    let mut manifest: Value = serde_json::from_str(manifest).expect("manifest JSON");
    manifest
        .as_object_mut()
        .expect("manifest object")
        .remove("manifestId")
        .expect("manifest identity");
    let decks = manifest["decks"].as_object_mut().expect("manifest decks");
    let north = decks.remove("north").expect("north deck");
    let south = decks.remove("south").expect("south deck");
    decks.insert("north".to_owned(), south);
    decks.insert("south".to_owned(), north);
    manifest["manifestId"] = json!(identity_hash(&manifest).expect("manifest identity"));
    canonical_json(&manifest).expect("canonical manifest")
}

fn manifest_with_first_seat(manifest: &str, first_seat: &str) -> String {
    let mut manifest: Value = serde_json::from_str(manifest).expect("manifest JSON");
    manifest
        .as_object_mut()
        .expect("manifest object")
        .remove("manifestId")
        .expect("manifest identity");
    manifest["firstSeat"] = json!(first_seat);
    manifest["manifestId"] = json!(identity_hash(&manifest).expect("manifest identity"));
    canonical_json(&manifest).expect("canonical manifest")
}

fn policy(manifest: &str, deck_id: &str) -> PolicySnapshot {
    let manifest: Value = serde_json::from_str(manifest).expect("manifest JSON");
    let mut body = json!({
        "authorityHash": manifest["authority"]["contentHash"],
        "deckId": deck_id,
        "engineVersion": "sorcery-core-v1",
        "generation": 0,
        "observationVersion": "seat-observation-v1",
        "schemaVersion": 1,
        "selector": {
            "atlasReserve": 3,
            "featurePriority": [
                "keep-mulligan", "play-site", "summon-minion", "preferred-draw",
                "powered-movement", "beneficial-tactic", "move-toward-enemy",
                "end-turn", "canonical-fallback"
            ]
        },
        "tieBreak": "canonical-action-order-v1"
    });
    body["policyId"] = json!(identity_hash(&body).expect("policy identity"));
    parse_policy_snapshot(&canonical_json(&body).expect("canonical policy")).expect("valid policy")
}

fn assert_invalid_pair_metadata<'a>(
    pair: GauntletPair<'a>,
    manifest: &'a str,
    swapped_manifest: &'a str,
    second_job: BatchJob<'a>,
) {
    let unswapped_second_job = BatchJob {
        manifest_json: manifest,
        ..second_job
    };
    let unswapped_pair = GauntletPair {
        orientations: [
            pair.orientations[0],
            GauntletOrientation {
                job: unswapped_second_job,
                ..pair.orientations[1]
            },
        ],
        ..pair
    };
    assert!(run_gauntlet(&[unswapped_pair], 1).is_err());

    let south_first_manifest = manifest_with_first_seat(swapped_manifest, "south");
    let south_first_pair = GauntletPair {
        orientations: [
            pair.orientations[0],
            GauntletOrientation {
                job: BatchJob {
                    manifest_json: &south_first_manifest,
                    ..second_job
                },
                ..pair.orientations[1]
            },
        ],
        ..pair
    };
    assert!(run_gauntlet(&[south_first_pair], 1).is_err());

    let mismatched_seed = GauntletPair { seed: 32, ..pair };
    assert!(run_gauntlet(&[mismatched_seed], 1).is_err());

    let different_pair = GauntletPair {
        orientations: [
            GauntletOrientation {
                north_deck_id: "deck-c",
                ..pair.orientations[0]
            },
            GauntletOrientation {
                south_deck_id: "deck-c",
                ..pair.orientations[1]
            },
        ],
        ..pair
    };
    assert!(run_gauntlet(&[pair, different_pair], 1).is_err());
}

fn assert_report_contract(report: &GauntletReport, value: &Value) {
    assert_eq!(report.games.len(), 2);
    assert_eq!(report.game_count, 2);
    assert_eq!(report.seeds, [31]);
    assert_eq!(report.by_seat.north.games, 2);
    assert_eq!(report.by_seat.south.games, 2);
    assert_eq!(report.by_deck["deck-a"].games, 2);
    assert_eq!(report.by_deck["deck-b"].games, 2);
    assert_eq!(
        value["bySeat"],
        json!({
            "north": { "draws": 0, "games": 2, "losses": 2, "wins": 0 },
            "south": { "draws": 0, "games": 2, "losses": 0, "wins": 2 },
        })
    );
    let expected_deck_counts = json!({
        "asNorth": { "draws": 0, "games": 1, "losses": 1, "wins": 0 },
        "asSouth": { "draws": 0, "games": 1, "losses": 0, "wins": 1 },
        "draws": 0,
        "games": 2,
        "losses": 1,
        "wins": 1,
    });
    assert_eq!(value["byDeck"]["deck-a"], expected_deck_counts);
    assert_eq!(value["byDeck"]["deck-b"], expected_deck_counts);
    assert!(
        report
            .games
            .iter()
            .all(|game| game.result.report.replay_verified)
    );
    assert_eq!(
        report
            .games
            .iter()
            .map(|game| {
                (
                    game.north_deck_id.as_str(),
                    game.seed,
                    game.south_deck_id.as_str(),
                )
            })
            .collect::<Vec<_>>(),
        [("deck-a", 31, "deck-b"), ("deck-b", 31, "deck-a"),]
    );
    let expected_average = report
        .games
        .iter()
        .map(|game| {
            u32::try_from(game.result.report.turn_count)
                .map(f64::from)
                .expect("fixture turn count fits u32")
        })
        .sum::<f64>()
        / 2.0;
    assert!((report.average_turns - expected_average).abs() < f64::EPSILON);
    assert_eq!(value["gameCount"], 2);
    assert_eq!(
        value["classification"],
        "unranked_partial_rules_unverified_authority"
    );
    assert_eq!(value["games"][0]["jobIndex"], 0);
    assert_eq!(value["games"][1]["jobIndex"], 1);
    assert!(value["games"][0].get("result").is_none());
    assert_eq!(value["games"][0]["report"]["replayVerified"], true);
    assert!(value["bySeat"].get("north").is_some());
    assert!(value["byDeck"]["deck-a"].get("total").is_none());
}

#[test]
fn seat_pairs_should_aggregate_identically_across_worker_counts() {
    let manifest = synthetic_demo_manifest_json(31).expect("seed-31 manifest");
    let swapped_manifest = swapped_manifest(&manifest);
    let baseline_policy = policy(&manifest, HASH_B);
    let alternate_policy = policy(&manifest, HASH_C);
    let first_job = BatchJob {
        manifest_json: &manifest,
        north_deck_id: baseline_policy.deck_id(),
        north_policy: &baseline_policy,
        south_deck_id: alternate_policy.deck_id(),
        south_policy: &alternate_policy,
    };
    let second_job = BatchJob {
        manifest_json: &swapped_manifest,
        north_deck_id: alternate_policy.deck_id(),
        north_policy: &alternate_policy,
        south_deck_id: baseline_policy.deck_id(),
        south_policy: &baseline_policy,
    };
    let pair = GauntletPair {
        seed: 31,
        orientations: [
            GauntletOrientation {
                job: first_job,
                north_deck_id: "deck-a",
                south_deck_id: "deck-b",
            },
            GauntletOrientation {
                job: second_job,
                north_deck_id: "deck-b",
                south_deck_id: "deck-a",
            },
        ],
    };

    let one = run_gauntlet(&[pair], 1).expect("one-worker gauntlet");
    let two = run_gauntlet(&[pair], 2).expect("two-worker gauntlet");
    let one_value = serde_json::to_value(&one).expect("serializable one-worker gauntlet");
    let two_value = serde_json::to_value(&two).expect("serializable two-worker gauntlet");

    assert_eq!(one, two);
    assert_eq!(
        serde_json::to_string(&one_value).expect("ordered one-worker gauntlet JSON"),
        serde_json::to_string(&two_value).expect("ordered two-worker gauntlet JSON")
    );
    assert_report_contract(&one, &one_value);

    assert_invalid_pair_metadata(pair, &manifest, &swapped_manifest, second_job);
}

File: src/simulator/gauntlet.ts
import { spawn } from 'node:child_process';
import { availableParallelism } from 'node:os';
import { fileURLToPath } from 'node:url';

import { canonicalJson, type JsonValue } from '../authority/canonical-json.ts';
import { identityHash } from '../authority/hash.ts';
import { deepFreeze, type EngineSeat } from '../engine/contract.ts';
import {
  createGameManifest,
  type GameDeckSpec,
  type GameManifest,
} from '../engine/game.ts';
import type { GameBatchResult } from '../commands/run-game-batch.ts';

const MAX_SEEDS = 128;
const MAX_WORKERS = 8;
const MAX_BATCH_BYTES = 64 * 1024 * 1024;
const MAX_OUTPUT_BYTES = 16 * 1024 * 1024;
const REPOSITORY_ROOT = fileURLToPath(new URL('../..', import.meta.url));

type OutcomeCounts = Readonly<{ draws: number; games: number; losses: number; wins: number }>;

export type GauntletDeck = Readonly<{ deck: GameDeckSpec; id: string }>;

export type TwoDeckGauntletInput = Readonly<{
  authority: GameManifest['authority'];
  cards: GameManifest['cards'];
  decks: readonly [GauntletDeck, GauntletDeck];
  seeds: readonly number[];
}>;

export type GauntletGameResult = Readonly<GameBatchResult & {
  northDeckId: string;
  seed: number;
  southDeckId: string;
}>;

export type TwoDeckGauntletReport = Readonly<{
  averageTurns: number;
  byDeck: Readonly<Record<string, Readonly<OutcomeCounts & {
    asNorth: OutcomeCounts;
    asSouth: OutcomeCounts;
  }>>>;
  bySeat: Readonly<Record<EngineSeat, OutcomeCounts>>;
  classification: 'unranked_partial_rules_unverified_authority';
  gameCount: number;
  games: readonly GauntletGameResult[];
  seeds: readonly number[];
}>;

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function countedRows(cardIds: readonly string[]): readonly Readonly<{
  cardId: string;
  copies: number;
}>[] {
  const counts = new Map<string, number>();
  cardIds.forEach((cardId) => counts.set(cardId, (counts.get(cardId) ?? 0) + 1));
  return [...counts]
    .sort(([left], [right]) => left < right ? -1 : left > right ? 1 : 0)
    .map(([cardId, copies]) => ({ cardId, copies }));
}

function deckCompositionId(deck: GameDeckSpec): ReturnType<typeof identityHash> {
  return identityHash({
    deck: {
      atlas: countedRows(deck.atlas),
      avatar: deck.avatar,
      spellbook: countedRows(deck.spellbook),
    },
    formatId: 'format:modeled-constructed-v1',
  } as JsonValue);
}

function policy(manifest: GameManifest, boundDeckId: ReturnType<typeof identityHash>): JsonValue {
  const body = {
    authorityHash: manifest.authority.contentHash,
    deckId: boundDeckId,
    engineVersion: manifest.engineVersion,
    generation: 0,
    observationVersion: 'seat-observation-v1',
    schemaVersion: 1,
    selector: {
      atlasReserve: 3,
      featurePriority: [
        'keep-mulligan', 'play-site', 'summon-minion', 'preferred-draw',
        'powered-movement', 'beneficial-tactic', 'move-toward-enemy',
        'end-turn', 'canonical-fallback',
      ],
    },
    tieBreak: 'canonical-action-order-v1',
  } as const;
  return { ...body, policyId: identityHash(body as unknown as JsonValue) } as JsonValue;
}

function jobPayload(manifest: GameManifest): JsonValue {
  const northDeckId = deckCompositionId(manifest.decks.north);
  const southDeckId = deckCompositionId(manifest.decks.south);
  return {
    manifestJson: canonicalJson(manifest as unknown as JsonValue),
    northDeckId,
    northPolicy: policy(manifest, northDeckId),
    southDeckId,
    southPolicy: policy(manifest, southDeckId),
  };
}

function rustRequest(input: TwoDeckGauntletInput, workers: number): string {
  const [deckA, deckB] = input.decks;
  return canonicalJson({
    deckAId: deckA.id,
    deckBId: deckB.id,
    pairs: input.seeds.map((seed) => ({
      aNorth: jobPayload(createGameManifest({
        authority: input.authority,
        cards: input.cards,
        decks: { north: deckA.deck, south: deckB.deck },
        firstSeat: 'north',
        seed,
      })),
      bNorth: jobPayload(createGameManifest({
        authority: input.authority,
        cards: input.cards,
        decks: { north: deckB.deck, south: deckA.deck },
        firstSeat: 'north',
        seed,
      })),
      seed,
    })),
    schemaVersion: 1,
    workers,
  } as JsonValue);
}

async function runRustGauntlet(request: string): Promise<unknown> {
  const child = spawn('cargo', [
    'run', '--release', '--locked', '--quiet', '-p', 'sorcery-engine',
    '--bin', 'sorcery-engine', '--', 'gauntlet-json',
  ], { cwd: REPOSITORY_ROOT, stdio: ['pipe', 'pipe', 'pipe'] });
  const stdout: Buffer[] = [];
  const stderr: Buffer[] = [];
  let stdoutBytes = 0;
  let stderrBytes = 0;
  return new Promise((resolvePromise, rejectPromise) => {
    let settled = false;
    const fail = (message: string): void => {
      if (settled) return;
      settled = true;
      child.kill();
      rejectPromise(new Error(message));
    };
    child.once('error', () => fail('gauntlet failed to start Rust'));
    child.stdin.once('error', () => fail('gauntlet failed while sending Rust input'));
    child.stdout.on('data', (chunk: Buffer) => {
      stdoutBytes += chunk.length;
      if (stdoutBytes > MAX_OUTPUT_BYTES) return fail('gauntlet Rust output exceeded its limit');
      stdout.push(chunk);
    });
    child.stderr.on('data', (chunk: Buffer) => {
      stderrBytes += chunk.length;
      if (stderrBytes > MAX_OUTPUT_BYTES) return fail('gauntlet Rust error output exceeded its limit');
      stderr.push(chunk);
    });
    child.once('close', (code) => {
      if (settled) return;
      if (code !== 0) {
        const unsupported = /not yet supported by Rust: ([A-Za-z][A-Za-z0-9:]*)/u
          .exec(Buffer.concat(stderr).toString('utf8'))?.[1];
        if (unsupported) {
          return fail(`gauntlet failed in Rust: unsupported fact ${unsupported}`);
        }
        return fail(`gauntlet failed in Rust: ${Buffer.concat(stderr).toString('utf8').trim() || 'unknown error'}`);
      }
      try {
        const text = Buffer.concat(stdout).toString('utf8').trim();
        const parsed: unknown = JSON.parse(text);
        if (canonicalJson(parsed as JsonValue) !== text) throw new Error('noncanonical output');
        settled = true;
        resolvePromise(parsed);
      } catch {
        fail('gauntlet Rust output was invalid');
      }
    });
    child.stdin.end(`${request}\n`);
  });
}

/** Runs a seat-swapped two-deck gauntlet inside the authoritative Rust batch path. */
export async function runTwoDeckGauntlet(
  input: TwoDeckGauntletInput,
  requestedWorkers = Math.min(availableParallelism(), MAX_WORKERS),
): Promise<TwoDeckGauntletReport> {
  const [deckA, deckB] = input.decks;
  if (!deckA.id.trim() || !deckB.id.trim() || deckA.id === deckB.id) {
    throw new RangeError('gauntlet deck IDs must be distinct nonempty strings');
  }
  if (input.seeds.length === 0 || input.seeds.length > MAX_SEEDS) {
    throw new RangeError('gauntlet must contain 1-128 seeds');
  }
  if (!Number.isSafeInteger(requestedWorkers)
    || requestedWorkers < 1
    || requestedWorkers > MAX_WORKERS) {
    throw new RangeError(`requestedWorkers must be 1-${MAX_WORKERS}`);
  }
  const request = rustRequest(input, requestedWorkers);
  if (Buffer.byteLength(request) > MAX_BATCH_BYTES) {
    throw new RangeError(`gauntlet exceeds ${MAX_BATCH_BYTES} bytes`);
  }
  const report = await runRustGauntlet(request);
  if (!isRecord(report)
    || report.classification !== 'unranked_partial_rules_unverified_authority'
    || report.gameCount !== input.seeds.length * 2
    || !Array.isArray(report.seeds)
    || report.seeds.length !== input.seeds.length
    || !Array.isArray(report.games)
    || report.games.length !== input.seeds.length * 2) {
    throw new Error('gauntlet results do not match the declared pairs');
  }
  return deepFreeze(report as unknown as TwoDeckGauntletReport);
}
